pkg/auth/authenticator/password/keystonepassword/keystonepassword.go

```diff
@@ -0,0 +1,84 @@
+package keystonepassword
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"runtime/debug"
+
+	"github.com/golang/glog"
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack"
+	"github.com/rackspace/gophercloud/openstack/identity/v3/tokens"
+
+	authapi "github.com/openshift/origin/pkg/auth/api"
+	"github.com/openshift/origin/pkg/auth/authenticator"
+	"k8s.io/kubernetes/pkg/auth/user"
+	"k8s.io/kubernetes/pkg/util"
+)
+
+// keystonePasswordAuthenticator uses OpenStack keystone to authenticate a user by password
+type keystonePasswordAuthenticator struct {
+	providerName   string
+	url            string
+	client         *http.Client
+	domainName     string
+	identityMapper authapi.UserIdentityMapper
+}
+
+// New creates a new password authenticator that uses OpenStack keystone to authenticate a user by password
+// A custom transport can be provided (typically to customize TLS options like trusted roots or present a client certificate).
+// If no transport is provided, http.DefaultTransport is used
+func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper) authenticator.Password {
+	if transport == nil {
+		transport = http.DefaultTransport
+	}
+	client := &http.Client{Transport: transport}
+	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper}
+}
+
+// AuthenticatePassword approves any login attempt which is successfully validated with Keystone
+func (a keystonePasswordAuthenticator) AuthenticatePassword(username, password string) (user.Info, bool, error) {
+	defer func() {
+		if e := recover(); e != nil {
+			util.HandleError(fmt.Errorf("Recovered panic: %v, %s", e, debug.Stack()))
+		}
+	}()
+
+	opts := gophercloud.AuthOptions{
+		IdentityEndpoint: a.url,
+		Username:         username,
+		Password:         password,
+		DomainName:       a.domainName,
+	}
+
+	// Calling NewClient/Authenticate manually rather than simply calling AuthenticatedClient
+	// in order to pass in a transport object that supports SSL
+	client, err := openstack.NewClient(opts.IdentityEndpoint)
+	client.HTTPClient = *a.client
+	if err == nil {
+		err = openstack.AuthenticateV3(client, opts)
+	}
+
+	if err != nil {
+		glog.Infof("Failed: Starting openstack authenticate client: %v", err)
+		if responseErr, ok := err.(*gophercloud.UnexpectedResponseCodeError); ok {
+			if responseErr.Actual == 401 {
+				return nil, false, nil
+			}
+		}
+		if err == tokens.ErrMissingPassword {
+			return nil, false, nil
+		}
+		return nil, false, errors.New("Failed to authenticate")
```

Comment:
I already have a logging statement above (first line in the block, above the 401 check). I can return the original error -- it gets swallowed by the UI code to display a generic "something went wrong" sort of error to the user anyway. Actually, come to think of it we probably don't need to log the error if it's a 401 response -- that's just a failed login, and if we get any other error here (other than the missing password one that we should handle differently), perhaps it should be a higher level log, since any error other than a 401 or a missing password probably indicates a config problem or a problem with keystone. Should it be glog.Errorf? (or Warningf)?